### PR TITLE
A comprehensive test suite

### LINKED ISSUE #72: Increase test coverage
At the moment there are functions in the API of Agents.jl that are not tested _directly_. Direct tests are useful, because they make bugs harder to hide, and during updates or enhancements they point out directly to the source of the problem.

Besides, increasing test coverage is an excellent way to contribute for a newcomer!

### PR DESCRIPTION
**Pull request by Libbum**
Closes #72 and #100 

This should bump our coverage up to 95% or above I'd guess. Fixed a number of small bugs and oversights in the process. 

Parallel cases have not been looked at, I'll tackle that at a later date. Maybe we should open a new issue specific to them.

### CODE REVIEW

**Comment by Libbum** [comment]
Will push a few more fixes up soon after seeing the coverage gaps.

**Comment by Libbum** [comment]
This is good to merge if you're happy with it.

File: src/core/agent_space_interaction.jl [hunk 1]
```diff
@@ -40,30 +40,39 @@ kill_agent!(id::Integer, model) = kill_agent!(model[id], model)
     genocide!(model::ABM)
 Kill all the agents of the model.
 """
-genocide!(model::ABM) =
+genocide!(model::ABM{A,<:DiscreteSpace}) where {A} = _genocide!(model)
+genocide!(model::ABM{A,Nothing}) where {A} = _genocide!(model)
```
Comment (Datseris) [inline src/core/agent_space_interaction.jl:44]: Both methods actually do the same thing, why dispatch on space here?

File: src/core/agent_space_interaction.jl [hunk 2]
```diff
@@ -40,30 +40,39 @@ kill_agent!(id::Integer, model) = kill_agent!(model[id], model)
...
     end
 
+
 """
     genocide!(model::ABM, n::Int)
 Kill the agents of the model whose IDs are larger than n.
 """
-function genocide!(model::ABM, n::Integer)
+genocide!(model::ABM{A,<:DiscreteSpace}, n::Integer) where {A} = _genocide!(model, n)
+genocide!(model::ABM{A,Nothing}, n::Integer) where {A} = _genocide!(model, n)
```
Comment (Datseris) [inline src/core/agent_space_interaction.jl:58]: Both methods actually do the same thing, why dispatch on space here?

File: src/core/agent_space_interaction.jl [hunk 3]
```diff
@@ -40,30 +40,39 @@ kill_agent!(id::Integer, model) = kill_agent!(model[id], model)
...
     end
-end
 
 """
     genocide!(model::ABM, f::Function)
 Kill all agents where the function `f(agent)` returns `true`.
 """
-function genocide!(model::ABM, f::Function)
+genocide!(model::ABM{A,<:DiscreteSpace}, f::Function) where {A} = _genocide!(model, f)
+genocide!(model::ABM{A,Nothing}, f::Function) where {A} = _genocide!(model, f)
```
Comment (Datseris) [inline src/core/agent_space_interaction.jl:70]: Both methods actually do the same thing, why dispatch on space here?

File: src/core/agent_space_interaction.jl (hunk 1 above)
Comment (Libbum) [reply src/core/agent_space_interaction.jl:44]: I kept getting ambiguous errors if we just set this as `genocide!(model)`. I fixed a bunch of continuous space issues after this though, so maybe those errors are dropped now. Give me a sec an I'll revert these to see...

Comment (Datseris) [reply src/core/agent_space_interaction.jl:44]: If it resolves ambiguity its okay.

Comment (Libbum) [reply src/core/agent_space_interaction.jl:44]: Yep. Here:
```julia
genocide!: Error During Test at /mnt/turtle/repos/Agents.jl/test/api_tests.jl:271
  Got exception outside of a @test
  MethodError: genocide!(::AgentBasedModel{Agent8,ContinuousSpace{typeof(Agents.defvel),Tuple{Int64,Int64}},typeof(fastest),Nothing}) is ambiguous. Candidates:
    genocide!(model::AgentBasedModel{A,#s60,F,P} where P where F where #s60<:ContinuousSpace) where A in Agents at /mnt/turtle/repos/Agents.jl/src/core/continuous_space.jl:214
    genocide!(model::AgentBasedModel) in Agents at /mnt/turtle/repos/Agents.jl/src/core/agent_space_interaction.jl:43
  Possible fix, define
    genocide!(::AgentBasedModel{A,S,F,P} where P where F where S<:ContinuousSpace) where A<:AbstractAgent
```

Call is:
```julia
  space2d = ContinuousSpace(2; periodic=true, extend=(1, 1))    
  model = ABM(Agent8, space2d)    
  attributes = (f1=true,f2=1)    
  for _ in 1:10 add_agent!(model, attributes...) end    
  genocide!(model)    
```
Ambiguous methods:
```julia
genocide!(model::ABM) = ... # agent_space_interaction.jl#43

genocide!(model::ABM{A,<:ContinuousSpace}) where {A} = ... # continuous_space.jl#214
```

How best to rectify that?

Comment (kavir1698) [reply src/core/agent_space_interaction.jl:44]: Maybe adding type info for genocide in discrete space?
```julia
# agent_space_interaction.jl#43
genocide!(model::ABM{A,<:DiscreteSpace}) where {A}
    for a in allagents(model)
        kill_agent!(a, model)
    end
end
```

Comment (Libbum) [reply src/core/agent_space_interaction.jl:44]: But then we don't have genocide for `Nothing`.

Comment (kavir1698) [reply src/core/agent_space_interaction.jl:44]: yeah, right!

On a different note, isn't massacre better than genocide?

Comment (Libbum) [reply src/core/agent_space_interaction.jl:44]:
```julia
genocide!(model)
decimate!(model) #where agent_ids.%10==0
massacre!(model) #where 20 < agent_ids < nagents(model)*0.1
```

Comment (Datseris) [reply src/core/agent_space_interaction.jl:44]: Let's minimize breaking changes by keeping `genocide!`. Other than that, I don't care as long as the name remains funny :D

**Pull request merged**